airflow/models/__init__.py

```diff
@@ -4012,17 +4088,55 @@ def create_dagrun(self,
             conf=conf,
             state=state
         )
+
+        dag_model = DagModel.get_dagmodel(self.dag_id)
+
+        tis = self.create_tis(execution_date)
+        edges = self.create_edges(-1)
+
+        last_run = dag_model.get_last_dagrun(include_externally_triggered=True)
+        if last_run is None:
+            same = False
+        else:
+            last_edges = DagEdge.fetch_edges_db(self.dag_id, last_run.graph_id)
+
+            # Compare edges from last run
+            same = True
+            if len(last_edges) != len(edges):
+                same = False
+            else:
+                e1 = set([(edge.from_task, edge.to_task) for edge in last_edges])
+                e2 = set([(edge.from_task, edge.to_task) for edge in edges])
+                if e1 != e2:
+                    same = False
+
+        if same:
+            # graph is not changed, keep last graph_id
+            graph_id = last_run.graph_id
+        elif last_run is None or last_run.graph_id is None:
+            # no graph known yet, beginning at 1
+            graph_id = 1
+        else:
+            # graph is changed
+            graph_id = last_run.graph_id + 1
+
+        for edge in edges:
+            edge.graph_id = graph_id
+
+        run.graph_id = graph_id
         session.add(run)
 
+        for ti in tis:
+            session.merge(ti)
+        if not same:
+            for edge in edges:
+                session.merge(edge)
```

Comment:
Since we know these are new objects, we can be easier on SQLA and let it not work out as much:

```suggestion
                session.add(edge)
```

Reply:
I would love this, only problem is that I tried this before and was hitting a ton of test errors. This means that all tests that are failing should clean the task_instanc, dag_edge and dag_run table. I will leave this open but not sure I should do this now, already touch to much tests in this PR ;)

Reply:
the `sync_to_db()` calls where not in there yet when I did test this first. I will try switching it to `.add` and see how much tests will fail. Maybe it's less as before because of all the changes already. It does help to make the test more stateless.

Reply:
I did try it and at least 37 tests are failing when I do this, maybe more but the log on travis was truncated because it was to big

Reply:
After fixing this 37 I'm getting 89 more failing tests :(. Now already working on so going to do an iterative testing

Reply:
One just left with 2 failing tests but this ones are not failing on my local machine again:
```
FAIL: test_scheduler_sla_miss_callback_exception (tests.test_jobs.SchedulerJobTest)
FAIL: test_scheduler_sla_miss_email_exception (tests.test_jobs.SchedulerJobTest)
```

Any idea how this sometimes can happen? It seems again have something to do with picking it seems.

Reply:
And finally all tests seems to work again :)